Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp

```diff
@@ -0,0 +1,209 @@
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "config.h"
+#include "CryptoAlgorithmEd25519.h"
+
+#if ENABLE(WEB_CRYPTO)
+
+#include "CryptoAlgorithmECDSA.h"
+#include "CryptoAlgorithmEcKeyParams.h"
+#include "CryptoAlgorithmEd25519Params.h"
+#include "CryptoKeyEC.h"
+#include "NotImplemented.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+#include <stdio.h>
+#include <string.h>
+#include <wtf/Assertions.h>
+#include <wtf/CrossThreadCopier.h>
+
+namespace WebCore {
+
+#if !PLATFORM(COCOA)
+ExceptionOr<Vector<uint8_t>> CryptoAlgorithmEd25519::platformSign(const CryptoKeyEC&, const Vector<uint8_t>&)
+{
+    ASSERT_NOT_REACHED();
+    notImplemented();
+    return Exception { NotSupportedError };
+}
+
+ExceptionOr<bool> CryptoAlgorithmEd25519::platformVerify(const CryptoKeyEC&, const Vector<uint8_t>&, const Vector<uint8_t>&)
+{
+    ASSERT_NOT_REACHED();
+    notImplemented();
+    return Exception { NotSupportedError };
+}
+
+#endif // PLATFORM(COCOA)
+
+Ref<CryptoAlgorithm> CryptoAlgorithmEd25519::create()
+{
+    return adoptRef(*new CryptoAlgorithmEd25519);
+}
+
+CryptoAlgorithmIdentifier CryptoAlgorithmEd25519::identifier() const
+{
+    return s_identifier;
+}
+
+void CryptoAlgorithmEd25519::generateKey(const CryptoAlgorithmParameters&, bool extractable, CryptoKeyUsageBitmap usages, KeyOrKeyPairCallback&& callback, ExceptionCallback&& exceptionCallback, ScriptExecutionContext&)
+{
+    
+    if (usages & (CryptoKeyUsageEncrypt | CryptoKeyUsageDecrypt | CryptoKeyUsageDeriveKey | CryptoKeyUsageDeriveBits | CryptoKeyUsageWrapKey | CryptoKeyUsageUnwrapKey)) {
+        exceptionCallback(SyntaxError);
+        return;
+    }
+    
+    auto result = CryptoKeyEC::generatePair(CryptoAlgorithmIdentifier::Ed25519, "Curve25519"_s, extractable, usages);
+    if (result.hasException()) {
+        exceptionCallback(result.releaseException().code());
```

Comment:
This does not need to be fixed right now, but it is a bit sad that we are dropping the message here.
Probably ExceptionCallback should take an Exception instead of an ExceptionCode.

Reply:
will file a bug for this task to be done on another patch